crates/ruff_dev/src/format_project.rs

```diff
@@ -0,0 +1,195 @@
+                    JaccardAccumulator::default()
+                }
+                Err(_) => {
+                    bar.println(format!("Formatting {} panicked", file.display()));
+                    JaccardAccumulator::default()
+                }
+            };
+            bar.inc(1);
+            Some(Ok((file, jaccard_accumulator)))
+        })
+        .collect::<anyhow::Result<Vec<_>>>()?;
+
+    let duration = start.elapsed();
+    bar.finish();
+
+    let jaccard_accumulator: JaccardAccumulator = files
+        .iter()
+        .map(|(_file, jaccard_accumulator)| *jaccard_accumulator)
+        .fold(JaccardAccumulator::default(), |x, y| x + y);
+
+    println!(
+        "Formatted {} files in {:.2}s with jaccard index {:.3}",
+        files.len(),
+        duration.as_secs_f32(),
+        jaccard_accumulator.jaccard_index(),
+    );
+    Ok(())
+}
+
+fn format_file(input_path: &Path) -> anyhow::Result<JaccardAccumulator> {
+    let content = fs::read_to_string(input_path).context("Failed to read file")?;
+    let printed = match format_module(&content, PyFormatOptions::default()) {
+        Ok(printed) => printed,
+        Err(err) => {
+            return if err
+                .to_string()
+                .starts_with("Source contains syntax errors ")
+            {
+                debug!(
+                    "Skipping {} with invalid first pass {}",
+                    input_path.display(),
+                    err
+                );
+                Ok(JaccardAccumulator::default())
+            } else {
+                Err(err)
+            };
+        }
+    };
+    let formatted = printed.as_code();
+
+    let reformatted = match format_module(formatted, PyFormatOptions::default()) {
+        Ok(reformatted) => reformatted,
+        Err(err) => {
+            bail!(
+                "Invalid syntax in formatted {}: {}",
+                input_path.display(),
+                err,
+            )
+        }
+    };
+
+    let mut jaccard_accumulator = JaccardAccumulator::default();
+    if content != reformatted.as_code() {
+        // Compute statistics
+        let diff = TextDiff::from_lines(content.as_str(), reformatted.as_code());
+        for change in diff.iter_all_changes() {
+            match change.tag() {
+                ChangeTag::Delete => jaccard_accumulator.black_input += 1,
+                ChangeTag::Insert => jaccard_accumulator.ruff_output += 1,
+                ChangeTag::Equal => jaccard_accumulator.intersection += 1,
+            }
+        }
+        // Simple atomic write.
+        // The file is in a directory so it must have a parent. Surprisingly `DirEntry` doesn't
+        // give us access without unwrap
+        let mut file = NamedTempFile::new_in(input_path.parent().unwrap())?;
+        file.write_all(reformatted.as_code().as_bytes())?;
+        // "If a file exists at the target path, persist will atomically replace it."
+        file.persist(input_path)?;
```

Comment:
Why does this file write any output? 

Rome had a useful script. It allowed you to generate a markdown report for all prettier test. The report contained:

* For files where the formatting differs
	* The Jacard index
	* The formatting diff
* The Jacard index for the whole project

I found this report extremely useful because:
* I had syntax highlighting of the patch
* I can see which files have the biggest differences
* I get an overview metric